File: src/main/java/com/iz/IzArchUnitPlugin.java
package com.iz;

import org.gradle.api.GradleException;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;

public class IzArchUnitPlugin implements Plugin<Project> {

    @Override
    public void apply(Project project) {

        IzArchUnitExtension archUnitExt = project.getExtensions()
                .create("archUnitTest", IzArchUnitExtension.class, project);

        IzArchUnitRuleTask applyRulesTask = project.getTasks().create("applyRules", IzArchUnitRuleTask.class);
        applyRulesTask.getArchUnitExtensions().convention(archUnitExt);

        final Task checkTask = findExistingTaskOrFailOtherwise("check", project);
        final Task testTask = findExistingTaskOrFailOtherwise("test", project);

        archUnitExt.setProject(project);

        checkTask.dependsOn(applyRulesTask);
        applyRulesTask.mustRunAfter(testTask);

    }

    private Task findExistingTaskOrFailOtherwise(String taskName, Project project){

        final Task taskToFind = project.getTasks().findByName(taskName);

        if (taskToFind==null){
            throw new GradleException("can't find the '"+taskName+"' task on which archUnitGradle task will depend - please check Gradle java plugin is applied");
        }

        return taskToFind;
    }
}

File: src/main/kotlin/com/iz/arunit/IzArchUnitPlugin.kt
package com.iz.arunit

import com.iz.arunit.IzArchUnitRuleTask
import org.gradle.api.GradleException
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.Task

class IzArchUnitPlugin : Plugin<Project> {

    override fun apply(project: Project) {
        val archUnitExt = project.extensions
            .create("izArchUnitTest", IzArchUnitExt::class.java, project)
        val applyRulesTask = project.tasks
            .create("applyIzArchRules", IzArchUnitRuleTask::class.java)

        val checkTask = findTaskOrThrowException("check", project)
        val testTask = findTaskOrThrowException("test", project)

        archUnitExt.project = project

        checkTask?.dependsOn(applyRulesTask)
        applyRulesTask.mustRunAfter(testTask)
    }

    private fun findTaskOrThrowException(taskName: String, project: Project): Task? {
        val task = project.tasks.findByName(taskName)
        return task?.let { it } ?:
            throw GradleException("No Task with name: '$task' on which iz-archunit task will depend")
    }
}